# Fragment Location Annotation Identification Mapper #

	 ___    _      _____  _         _____  ___    ___    ___    ___   
	(  _`\ ( )    (  _  )(_)/'\_/`\(  _  )(  _`\ (  _`\ (  _`\ |  _`\ 
	| (_(_)| |    | (_) || ||     || (_) || |_) )| |_) )| (_(_)| (_) )
	|  _)  | |  _ |  _  || || (_) ||  _  || ,__/'| ,__/'|  _)_ | ,  / 
	| |    | |_( )| | | || || | | || | | || |    | |    | (_( )| |\ \ 
	(_)    (____/'(_) (_)(_)(_) (_)(_) (_)(_)    (_)    (____/'(_) (_)

### Introduction ###
FlaiMapper follows a 5-step workflow to annotate dncRNAs by computing the position of 5’-and 3’-ends of candidate dncRNAs:

1. Read FlaiMapper input
2. Determine 4 starting data sets for each subject ncRNA.
3. Identify candidate dncRNAs by peak detection of possible dncRNA ends (start and stop positions)
4. Smooth filtering to extract the most probable dncRNA ends from step 3 by an arbitrary cut-off
5. Back Tracing to identify the start and stop positions which are most likely to belong to the same dncRNA

### Installation ###

### Usage ###
#### Reproducing the published experiment ####
All documents used for the publication can be generated by running the scripts
in the '[../scripts](https://github.com/yhoogstrate/flaimapper/tree/master/scripts/)' directory.

#### FlaiMapper ####
To perform a FlaiMapper analysis ensure the files that belong to a sample have the following directory structure:

	.
	├── idreadable.txt
	└── validated
	    ├── file100.fa
	    ├── file101.fa
	    ├── file102.fa
	    ├── file103.fa
	    ├── file104.fa
	    ├── file105.fa
	    └── file106.fa

The _idreable.txt_, that links the fasta files to a reference ncRNA sequence, must be a tab-delimted file of the following syntax:

	sequence	filename
	>NAME=TRNAValAAC&LOCI=[chr1:180184276-180184348:strand=-]&SOURCE=UCSC&SOURCE-ACCESSION=chr1.tRNA63-ValAAC&GENOME=hg19	file100
	>HGNC=38236&HUGO-Symbol=MIR3198-1&HUGO-Name=microRNA_3198-1&LOCI=[chr22:18246936-18247035:strand=-]&SOURCE=RefSeq&SOURCE-ACCESSION=NR_036168&GENOME=hg19	file101

The individual fasta (_*.fa_) files, located in a directory valled "**validated**", must be of the following sytax:

	>NAME=TRNAVALAAC&LOCI=[CHR1:180184276-180184348:STRAND=-]&SOURCE=UCSC&SOURCE-ACCESSION=CHR1.TRNA63-VALAAC&GENOME=HG19
	GTTTCCATAGTGTACTGGTTATCACATTCACCTAACACGCGAAAGGTCCTTGGTTTGAAACCAGGCAGAAACACCA
	>SOLEXA_READ_1_OF_10_GTTTCCATAGTGTAGTGGTTATC_HITS8
	GTTTCCATAGTGTAGTGGTTATC-----------------------------------------------------
	>SOLEXA_READ_1_OF_12_GTTTCCATAGTGTAGTGG_HITS7
	GTTTCCATAGTGTAGTGG----------------------------------------------------------
	>SOLEXA_READ_1_OF_11_GTTTCCATAGTGTAGTGGTTAT_HITS6
	GTTTCCATAGTGTAGTGGTTAT------------------------------------------------------

Then FlaiMapper is executed as follows:

	flaimapper.py [-h] [-v | -q] [-o OUTPUT] [-f FORMAT]
	                     alignment_directories [alignment_directories ...]
	
	positional arguments:
	  alignment_directories
	
	optional arguments:
	  -h, --help            show this help message and exit
	  -v, --verbose
	  -q, --quiet
	  -o OUTPUT, --output OUTPUT
	                        output filename; '-' for stdout
	  -f FORMAT, --format FORMAT
	                        file format of the output: [1: table; per fragment],
	                        [2: table; per ncRNA], [3: genbank]

Given that we have two samples ('[SRR038853](https://github.com/yhoogstrate/flaimapper/tree/master/share/small_RNA-seq_alignments/SRP002175/SRR038853)' and '[SRR038855](https://github.com/yhoogstrate/flaimapper/tree/master/share/small_RNA-seq_alignments/SRP002175/SRR038855)') of experiment '[SRP002175](https://github.com/yhoogstrate/flaimapper/tree/master/share/small_RNA-seq_alignments/SRP002175)' of which we would like to combine the data, we run FlaiMapper with:

	../src/flaimapper.py \
	                     -v \
	                     -o ../output/SRP002175/01_output_flaimapper.txt \
	                     -f 1 \
	                        ../share/small_RNA-seq_alignments/SRP002175/SRR038853 \
	                        ../share/small_RNA-seq_alignments/SRP002175/SRR038855 \

This predictes the ncRNA fragments using the combination of the data and puts the "_table; per fragment_" type of output in directory '[../output/SRP002175/01_output_flaimapper.txt](https://github.com/yhoogstrate/flaimapper/output/SRP002175/01_output_flaimapper.txt)'.

### Authors ###
The people who have contributed to this project are:

 - Youri Hoogstrate
 - Elena S. Martens-Uzunova
 - Guido Jenster

### Citing ###

